OpenTrons/opentrons-api
# refactor(robot-server): add basic SQL migrations to persistence layer

## Overview

This PR adds a basic SQL migration system to the robot-server's protocol/run persistence layer. Since the database has changed slightly since the 5.1.0-beta.0 release, **this PR is necessary for the robot server to continue functioning** if you've installed that version or any copy of `edge` from after SQL was added but before 2022-05-17.

#10187 was merged, and it (intentionally) included some database schema changes. This left us with a few options to make sure we didn't break the robot server for too long:

1. Force anyone with a previous (unreleased to prod) version of the SQL database to wipe it out before using the latest version
    - Ruled this out because #10238 is not yet completed, and `5.1.0-beta.0` is live
2. Make edits to #10187 to not modify the existing tables
    - Ruled this would've made the persistence logic less robust
3. Add a migration system to #10187 
    - Ruled this out because the PR was massive and we'd already been through a big review cycle
4. Merge #10187, then add a migration system in a new PR
    - AKA this PR

## Changelog

- add basic SQL migrations to persistence layer
- tweak a column name to better match what we came up with during review of #10187, since we're in here adding migrations anyway

## Review requests

Hopefully the code is at least a little self-evident, but I;ll be working on spinning up some docs about this system. This PR adds a `migration` table, that is used track any time the SQL database is migrated to a new version. New versions are necessitated by column changes to existing tables. These migration capabilities were driven entirely by what was easy / exactly what we need for #10187.

| change | needs migration | notes
| --- | --- | --- |
| new table | no | SQLAlchemy will add the table if it's not there |
| new column | yes | Must be nullable or have a default value |
| change column type | yes, but | **We can't / shouldn't do this** |
| drop column | yes, but | **We can't / shouldn't do this** |
| drop table | yes, but | **This is a pretty bad idea** |

In the past, we talked about using Alembic for SQL migrations, and I hope this PR doesn't close that door if we ever need more sophisticated migration capabilities. For where we're at, though, (and after talking with @TamarZanzouri), Alembic felt really heavy weight for our needs.

Instead, this PR implements a really lightweight, up-migration-only system that simply executes raw SQL strings to add columns. It gets the job done, especially considering SQLAlchemy offers no `ALTER TABLE` API that I could find.

Please also look at the unit tests and let me know what you think.

### How migration works

Migration is run at SQLAlchemy `Engine` creation, after tables have been added.

1. Find the current version 
    - The table is inspected to determine if it's at version 0 (shape prior to #10187, missing some `run` columns
    - If it's not 0, the version number is pulled from the `migration` table
2. If the found version is less than 1, the function `_migrate_0_to_1` is called
    - The idea being we tack on more migration functions in the future
3. If any migrations were run, a new entry with the latest version is added to the `migration` record to tell use "DB was migrated to version {version} at {datetime}"

### Testing

1. Load `5.1.0-beta.0` onto a robot
2. `POST` some `/protocols` and `/runs` to get the database initialized with some data
3. `make push` this branch (or load a CI build via `ot2-system.zip`)
4. `GET /protocols` and `/runs` and verify that requests succeed and contain migrated data
5. Ensure you can still `POST` new resources

## Risk assessment

Pretty low I think. I've smoke tested this on my dev machine with a database inspector and on actual hardware that had data persisted with the previous schema. The old runs were able to migrate just fine.
